sycl/source/detail/kernel_program_cache.hpp

```diff
@@ -427,23 +463,31 @@ class KernelProgramCache {
     return std::make_pair(It->second, DidInsert);
   }
 
-  template <typename KeyT>
-  KernelFastCacheValT tryToGetKernelFast(KeyT &&CacheKey) {
-    KernelFastCacheReadLockT Lock(MKernelFastCacheMutex);
-    auto It = MKernelFastCache.find(CacheKey);
-    if (It != MKernelFastCache.end()) {
-      traceKernel("Kernel fetched.", CacheKey.second, true);
+  FastKernelCacheValT
+  tryToGetKernelFast(KernelNameStrRefT KernelName, ur_device_handle_t Device,
+                     FastKernelSubcacheT *KernelSubcacheHint) {
+    KernelFastCacheReadLockT Lock(MFastKernelCacheMutex);
```

Comment:
Hmm, I think there is a potential data race if the same `KernelSubcacheHint` is used by two different contexts, then protecting with the `MFastKernelCacheMutex` is not enough. So the `KernelFastCacheReadLockT Lock(MFastKernelCacheMutex);` should be inside the `if` block at L470. And to protect access to the `KernelSubcacheHint` we need a separate mutex which is part of the `FastKernelSubcacheT` structure.

Reply:
Yeah, I think you're right, this needs to be changed. Good catch!